spec/java/util/HashMap.Values.lsl

```diff
@@ -0,0 +1,535 @@
+    }
+
+
+    @AutoInline @Phantom proc _throwNPE (): void
+    {
+        action THROW_NEW("java.lang.NullPointerException", []);
+    }
+
+
+    proc _mapToValuesArray (): array<Object>
+    {
+        val storageSize: int = action MAP_SIZE(this.storage);
+        result = action ARRAY_NEW("java.lang.Object", storageSize);
+        val unseen: map<Object, Map_Entry<Object, Object>> = action MAP_CLONE(this.storage);
+        var i: int = 0;
+        action LOOP_FOR(
+            i, 0, storageSize, +1,
+            _mapToValuesArray_loop(i, result, unseen)
+        );
+    }
+
+
+    @Phantom proc _mapToValuesArray_loop (i: int, result: array<Object>, unseen: map<Object, Map_Entry<Object, Object>>): void
+    {
+        val curKey: Object = action MAP_GET_ANY_KEY(unseen);
+        val entry: Map_Entry<Object, Object> = action MAP_GET(unseen, curKey);
+        result[i] = action CALL_METHOD(entry, "getValue", []);
+        action MAP_REMOVE(unseen, curKey);
+    }
+
+
+    // constructors
+
+    @private constructor *.`<init>` (@target self: HashMap_Values, _this: HashMap)
+    {
+        // #note: default constructor without any body, like in the original class
+    }
+
+
+    // static methods
+
+    // methods
+
+    // within java.util.AbstractCollection
+    fun *.add (@target self: HashMap_Values, e: Object): boolean
+    {
+        if (true) // <- fooling Java compiler to not give "unreachable statement" error
+            _throwUOE();
+    }
+
+
+    // within java.util.AbstractCollection
+    fun *.addAll (@target self: HashMap_Values, c: Collection): boolean
+    {
+        if (true) // <- fooling Java compiler to not give "unreachable statement" error
+            _throwUOE();
+    }
+
+
+    @final fun *.clear (@target self: HashMap_Values): void
+    {
+        HashMapAutomaton(this.parent).modCount += 1;
+        this.storage = action MAP_NEW();
+    }
+
+
+    @final fun *.contains (@target self: HashMap_Values, value: Object): boolean
+    {
+        result = false;
+        val storageSize: int = action MAP_SIZE(this.storage);
+        if (storageSize != 0)
+        {
+            val unseen: map<Object, Map_Entry<Object, Object>> = action MAP_CLONE(this.storage);
+            var i: int = 0;
+            action LOOP_WHILE(
+                result != true,
+                _containsValue_loop(result, unseen, value)
+            );
+        }
+    }
+
+
+    @Phantom proc _containsValue_loop (result: boolean, unseen: map<Object, Map_Entry<Object, Object>>, value: Object): void
+    {
+        val curKey: Object = action MAP_GET_ANY_KEY(unseen);
+        val entry: Map_Entry<Object, Object> = action MAP_GET(unseen, curKey);
+        val curValue: Object = action CALL_METHOD(entry, "getValue", []);
+        if (action OBJECT_EQUALS(curValue, value))
+            result = true;
+        else
+            action MAP_REMOVE(unseen, curKey);
+    }
+
+
+    // #note: double loop... Can we avoid this ? too comprehensive realization...
```

Comment:
@dog-m 
I have made little experiment:
```
        HashMap<Object, Object> map1 = new HashMap<>();
        map1.put("A", 1);
        map1.put("B", 1);
        map1.put("C", 2);
        System.out.println(map1.values().containsAll(Arrays.asList(1, 1, 1)));
```
// result = true*

Maybe collect all values from our storage to newMap (where key will be - values from storageRef); O(n1)
Collect values from Collection to secondMap; O(n2)

Make cycle with checking MAP_HAS_KEY; 
*if n2>n1 => O(n1); else O(n2);*

What do you think about this idea ?

Reply:
Your suggestion has at least two downsides:
1. On every presence check there **will** be calls to `hashCode` method if we use `MAP_NEW` action (on the other hand, `IDENTITY_MAP_NEW` would not make such calls, instead checking just equality between references).
2. Keys cannot be `null` for a map produced by `MAP_NEW` action (`key.hashCode()` will error-out). This forces us to specifically check for `null` values (by setting a flag or something that indicates that there are `null`s).

`IDENTITY_MAP_NEW` also would not help much us here. We **need** to somehow check the equality between objects via corresponding *user-controlled* `equals` method.
I don't see a better way here.

I see that this can be seen similar to `Stream.distinct()` method, but it is not in my opinion.

Reply:
<details>
<summary>Minor correction to my original suggestion:</summary>

```dart
@Phantom proc containsAll_inspector (iter: Iterator<Object>, thisValues: array<Object>, thisSize: int, i: int, result: boolean): void
{
    val value: Object = action CALL_METHOD(iter, "next", []);

    result = false;
    i = 0;
    action LOOP_WHILE(
        result == false && i != thisSize,
        containsAll_inspector_loop(i, value, thisValues, result)
    );
}

@Phantom proc containsAll_inspector_loop (i: int, value: Object, thisValues: array<Object>, result: boolean): void
{
    result = action OBJECT_EQUALS(thisValues[i], value);

    i += 1;
}
```

</details>